Mapbox snapshot API causes ANR when used in recycler view

Mapbox snapshot API causes ANR when used in recycler view.

Map usage : Each item in the recycler view is a map. After map is rendered on the screen, Screenshot API(MapboxMap.snapshot()) is used to capture the snapshot of the map. However, an ANR is observed. Attached Stack trace below.

Platform: Android 29
Mapbox SDK version: 9.0.0 

Steps to trigger behavior (Reproduction rate - 8/10)

 1. Create a recycler view with map as each item.
 2. Set Style, add source and layers to the maps.
 3. Once fully rendered, capture the screenshot of the maps using screenshot API 
 4. Scroll the recycler view
 5. ANR is observed

Expected behavior

1. ANR should not happen

Actual behavior
	
1.	ANR observed 

> --------- beginning of crash
05-08 10:01:38.343 13348 13459 E AndroidRuntime: FATAL EXCEPTION: |ANR-WatchDog|
05-08 10:01:38.343 13348 13459 E AndroidRuntime: Process: com.splunk.android.alerts.debug, PID: 13348
05-08 10:01:38.343 13348 13459 E AndroidRuntime: com.github.anrwatchdog.ANRError: Application Not Responding for at least 5000 ms.
05-08 10:01:38.343 13348 13459 E AndroidRuntime: Caused by: com.github.anrwatchdog.ANRError$$$_Thread: main (state = RUNNABLE)
05-08 10:01:38.343 13348 13459 E AndroidRuntime: 	at com.mapbox.mapboxsdk.maps.renderer.MapRenderer.nativeReset(Native Method)
05-08 10:01:38.343 13348 13459 E AndroidRuntime: 	at com.mapbox.mapboxsdk.maps.renderer.glsurfaceview.GLSurfaceViewMapRenderer.access$000(GLSurfaceViewMapRenderer.java:21)
05-08 10:01:38.343 13348 13459 E AndroidRuntime: 	at com.mapbox.mapboxsdk.maps.renderer.glsurfaceview.GLSurfaceViewMapRenderer$1.onGLSurfaceViewDetached(GLSurfaceViewMapRenderer.java:43)
05-08 10:01:38.343 13348 13459 E AndroidRuntime: 	at com.mapbox.mapboxsdk.maps.renderer.glsurfaceview.MapboxGLSurfaceView.onDetachedFromWindow(MapboxGLSurfaceView.java:27)
05-08 10:01:38.343 13348 13459 E AndroidRuntime: 	at android.view.View.dispatchDetachedFromWindow(View.java:20160)
05-08 10:01:38.343 13348 13459 E AndroidRuntime: 	at android.view.ViewGroup.dispatchDetachedFromWindow(ViewGroup.java:3850)
05-08 10:01:38.343 13348 13459 E AndroidRuntime: 	at android.view.ViewGroup.dispatchDetachedFromWindow(ViewGroup.java:3850)
05-08 10:01:38.343 13348 13459 E AndroidRuntime: 	at android.view.ViewGroup.dispatchDetachedFromWindow(ViewGroup.java:3850)
05-08 10:01:38.343 13348 13459 E AndroidRuntime: 	at android.view.ViewGroup.dispatchDetachedFromWindow(ViewGroup.java:3850)
05-08 10:01:38.343 13348 13459 E AndroidRuntime: 	at android.view.ViewGroup.dispatchDetachedFromWindow(ViewGroup.java:3850)
05-08 10:01:38.343 13348 13459 E AndroidRuntime: 	at android.view.ViewGroup.dispatchDetachedFromWindow(ViewGroup.java:3850)
05-08 10:01:38.343 13348 13459 E AndroidRuntime: 	at android.view.ViewGroup.dispatchDetachedFromWindow(ViewGroup.java:3850)
05-08 10:01:38.343 13348 13459 E AndroidRuntime: 	at android.view.ViewGroup.removeViewInternal(ViewGroup.java:5475)
05-08 10:01:38.343 13348 13459 E AndroidRuntime: 	at android.view.ViewGroup.removeViewAt(ViewGroup.java:5422)
05-08 10:01:38.343 13348 13459 E AndroidRuntime: 	at androidx.recyclerview.widget.RecyclerView$5.removeViewAt(RecyclerView.java:882)
05-08 10:01:38.343 13348 13459 E AndroidRuntime: 	at androidx.recyclerview.widget.ChildHelper.removeViewAt(ChildHelper.java:168)
05-08 10:01:38.343 13348 13459 E AndroidRuntime: 	at androidx.recyclerview.widget.RecyclerView$LayoutManager.removeViewAt(RecyclerView.java:8639)
05-08 10:01:38.343 13348 13459 E AndroidRuntime: 	at androidx.recyclerview.widget.RecyclerView$LayoutManager.removeAndRecycleViewAt(RecyclerView.java:8912)
05-08 10:01:38.343 13348 13459 E AndroidRuntime: 	at androidx.recyclerview.widget.LinearLayoutManager.recycleChildren(LinearLayoutManager.java:1429)
05-08 10:01:38.343 13348 13459 E AndroidRuntime: 	at androidx.recyclerview.widget.LinearLayoutManager.recycleViewsFromStart(LinearLayoutManager.java:1479)
05-08 10:01:38.343 13348 13459 E AndroidRuntime: 	at androidx.recyclerview.widget.LinearLayoutManager.recycleByLayoutState(LinearLayoutManager.java:1554)
05-08 10:01:38.343 13348 13459 E AndroidRuntime: 	at androidx.recyclerview.widget.LinearLayoutManager.fill(LinearLayoutManager.java:1578)
05-08 10:01:38.343 13348 13459 E AndroidRuntime: 	at androidx.recyclerview.widget.LinearLayoutManager.scrollBy(LinearLayoutManager.java:1391)
05-08 10:01:38.343 13348 13459 E AndroidRuntime: 	at androidx.recyclerview.widget.LinearLayoutManager.scrollVerticallyBy(LinearLayoutManager.java:1128)
05-08 10:01:38.343 13348 13459 E AndroidRuntime: 	at androidx.recyclerview.widget.RecyclerView.scrollStep(RecyclerView.java:1841)
05-08 10:01:38.343 13348 13459 E AndroidRuntime: 	at androidx.recyclerview.widget.RecyclerView$ViewFlinger.run(RecyclerView.java:5302)
05-08 10:01:38.343 13348 13459 E AndroidRuntime: 	at android.view.Choreographer$CallbackRecord.run(Choreographer.java:967)
05-08 10:01:38.343 13348 13459 E AndroidRuntime: 	at android.view.Choreographer.doCallbacks(Choreographer.java:791)
05-08 10:01:38.343 13348 13459 E AndroidRuntime: 	at android.view.Choreographer.doFrame(Choreographer.java:722)
05-08 10:01:38.343 13348 13459 E AndroidRuntime: 	at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:952)
05-08 10:01:38.343 13348 13459 E AndroidRuntime: 	at android.os.Handler.handleCallback(Handler.java:883)
05-08 10:01:38.343 13348 13459 E AndroidRuntime: 	at android.os.Handler.dispatchMessage(Handler.java:100)
05-08 10:01:38.343 13348 13459 E AndroidRuntime: 	at android.os.Looper.loop(Looper.java:214)
05-08 10:01:38.343 13348 13459 E AndroidRuntime: 	at android.app.ActivityThread.main(ActivityThread.java:7356)
05-08 10:01:38.343 13348 13459 E AndroidRuntime: 	at java.lang.reflect.Method.invoke(Native Method)
05-08 10:01:38.343 13348 13459 E AndroidRuntime: 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492)
05-08 10:01:38.343 13348 13459 E AndroidRuntime: 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:930)

--- a/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/maps/MapboxMap.java
+++ b/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/maps/MapboxMap.java
@@ -81,6 +81,7 @@ public final class MapboxMap {
   private Style style;
 
   private boolean debugActive;
+  private boolean started;
 
   MapboxMap(NativeMap map, Transform transform, UiSettings ui, Projection projection,
             OnGesturesManagerInteractionListener listener, CameraChangeDispatcher cameraChangeDispatcher,
@@ -143,6 +144,7 @@ public final class MapboxMap {
    * Called when the hosting Activity/Fragment onStart() method is called.
    */
   void onStart() {
+    started = true;
     locationComponent.onStart();
   }
 
@@ -150,6 +152,7 @@ public final class MapboxMap {
    * Called when the hosting Activity/Fragment onStop() method is called.
    */
   void onStop() {
+    started = false;
     locationComponent.onStop();
   }
 
@@ -2003,6 +2006,9 @@ public final class MapboxMap {
    * @param callback Callback method invoked when the snapshot is taken.
    */
   public void snapshot(@NonNull SnapshotReadyCallback callback) {
+    if (!started) {
+      return;
+    }
     nativeMapView.addSnapshotCallback(callback);
   }
 

--- a/MapboxGLAndroidSDKTestApp/src/main/java/com/mapbox/mapboxsdk/testapp/activity/maplayout/GLSurfaceRecyclerViewActivity.kt
+++ b/MapboxGLAndroidSDKTestApp/src/main/java/com/mapbox/mapboxsdk/testapp/activity/maplayout/GLSurfaceRecyclerViewActivity.kt
@@ -6,7 +6,6 @@ import android.view.LayoutInflater
 import android.view.ViewGroup
 import android.widget.TextView
 import androidx.appcompat.app.AppCompatActivity
-import androidx.recyclerview.widget.RecyclerView
 import com.mapbox.mapboxsdk.maps.MapView
 import com.mapbox.mapboxsdk.maps.Style
 import com.mapbox.mapboxsdk.testapp.R
@@ -141,7 +140,11 @@ open class GLSurfaceRecyclerViewActivity : AppCompatActivity() {
       }
 
       fun bind(mapItem: MapItem) {
-        mapView.getMapAsync { mapboxMap -> mapboxMap.setStyle(mapItem.style) }
+        mapView.getMapAsync { mapboxMap ->
+          mapboxMap.setStyle(mapItem.style) {
+            mapboxMap.snapshot { }
+          }
+        }
       }
     }
 

